### PR TITLE
Fixed custom entity selectors

### LINKED ISSUE #3558: Custom EntitySelector predicates are broken [1.11+]
Minecraft 1.11 introduced a change in the handling of unknown selector arguments:
```JAVA
//Snippet from EntitySelector.getArgumentMap
if (!IS_VALID_ARGUMENT.apply(name))
{
    throw new CommandException("commands.generic.selector_argument", new Object[] {s});
}
```
`IS_VALID_ARGUMENT.apply` does the following:
```JAVA
 return name != null && (EntitySelector.VALID_ARGUMENTS.contains(name) || name.length() > "score_".length() && name.startsWith("score_"));
```

This means that the current event based solution for handling custom predicates (introduced by #3356) should no longer work, as a list of known parameter names has to exist beforehand.

One possible solution would be register a parameter name together with a `Predicate<Entity>`-factory that creates the predicate.

### PR DESCRIPTION
**Pull request by maxanier**
Fixes #3558.
As MC 1.11 requires any entity selector to be registered beforehand a change was required.
The current event based solution seemed a bit odd as it would require two independent actions (registering the selector string at startup and creating an event handler), so this PR switches to a factory based way (modders have to register a factory together with the expected selector strings at startup).
Mods (which already  offer custom selectors) have to adapt, but since the event based solution would have required changes as well and switching to the factory method is not very time-consuming, this should be no problem. 
To keep binary compatibility with existent 1.11 mods, the old event still exists, but is deprecated and not fired.

I was not sure where to put the code, so if you have a better location, I will move it.

This also adds a test mod, which registers a 'forge:min_health' selector.
"/kill @e[forge:min_health=20]" should only kill entities with 20 or more health.

### CODE REVIEW

**Comment by maxanier** [comment]
@Mathe172 This should work with your #3545, doesn't it. Should we somehow merge both PRs? I could move the factory registrations to your SelectorHandlerManager or similar. Or should we keep the two systems separate as they do slightly different things (and to increase the chance that your PR is merged anytime soon as you are working on it for some time now).

**Comment by Mathe172** [comment]
I think the two PRs are too different to be merged in a meaningful way - this one upgrades the vanilla selector, while the other one is for defining new selectors.
Nevertheless, they should work together nicely - a mod defining a new selector could even call `createEntitySelectors` and everything should still work

File: src/test/java/net/minecraftforge/test/CustomEntitySelectorTest.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+                    final int value = Integer.parseInt(health);
+                    predicates.add(new Predicate<Entity>()
+                    {
+                        @Override
+                        public boolean apply(@Nullable Entity input)
+                        {
+                            if (!(input instanceof EntityLivingBase))
+                                return false;
+                            boolean health = ((EntityLivingBase) input).getHealth() >= value;
+                            return invert && !health || !invert && health;
```
Comment (Mathe172) [inline src/test/java/net/minecraftforge/test/CustomEntitySelectorTest.java]: Could be changed to `invert != health`

File: src/test/java/net/minecraftforge/test/CustomEntitySelectorTest.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+                final boolean invert = health.startsWith("!");
+
+                if (invert)
+                {
+                    health = health.substring(1);
+                }
+                try
+                {
+                    final int value = Integer.parseInt(health);
+                    predicates.add(new Predicate<Entity>()
```
Comment (Mathe172) [inline src/test/java/net/minecraftforge/test/CustomEntitySelectorTest.java]: You could just return `Collections.singletonList(...)`

File: src/test/java/net/minecraftforge/test/CustomEntitySelectorTest.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+    private class EntitySelectorFactory implements IEntitySelectorFactory
+    {
+
+        @Nonnull @Override
+        public List<Predicate<Entity>> createPredicates(Map<String, String> arguments, String mainSelector, ICommandSender sender, Vec3d position)
+        {
+            List<Predicate<Entity>> predicates = Lists.newArrayList();
+            String health = arguments.get("forge:min_health");
+            if (health != null)
```
Comment (Mathe172) [inline src/test/java/net/minecraftforge/test/CustomEntitySelectorTest.java]: Probably easier to just return `Collections.emptyList()`

**Review by Mathe172 — COMMENTED** [review]
Since they apply to just a test mod, the suggested changes are probably not very important - they're only there for the sake of completeness :)

File: src/test/java/net/minecraftforge/test/CustomEntitySelectorTest.java (hunk 3 above)
Comment (maxanier) [reply src/test/java/net/minecraftforge/test/CustomEntitySelectorTest.java]: Vanilla sometimes uses my approach, sometimes yours. I decided to go with this as I personally think it's easier to understand. 
But as modders might take this as an example and as every (tiny) performance increase counts, I changed it according to your suggestions now :)

File: src/test/java/net/minecraftforge/test/CustomEntitySelectorTest.java (hunk 1 above)
Comment (maxanier) [reply src/test/java/net/minecraftforge/test/CustomEntitySelectorTest.java]: :man_facepalming:

**Comment by maxanier** [comment]
Any other concerns about this, or can this be merged? @mezz

**Comment by mezz** [comment]
I think this requires #3545 to be merged first, so I haven't reviewed it yet.

**Comment by maxanier** [comment]
@mezz can you take a look at this again?
I don't think this requires #3545, they affect similar things, but not the same. This one allows custom selectors "[forge:min_health=5]" to be defined while the other one allows custom prefixes (something like "@k[...]") to be defined.
As far as I can tell they should work together just fine.

The old (not working way) is in the code since beginning of 1.11, so modders might start wondering why this does not work, and I finally got around updating my mods to 1.11 and would like to use it as well :)

File: src/main/java/net/minecraftforge/event/ForgeEventFactory.java [hunk 4]
```diff
@@ -253,13 +253,6 @@ public static String getPlayerDisplayName(EntityPlayer player, String username)
         return event.getDisplayname();
     }
 
-    public static List<Predicate<Entity>> gatherEntitySelectors(Map<String, String> map, String mainSelector, ICommandSender sender, Vec3d position)
```
Comment (mezz) [inline src/main/java/net/minecraftforge/event/ForgeEventFactory.java]: To preserve binary compatibility, deprecate this and leave the code in for now.
Leave a javadoc comment saying it should be removed in 1.12

File: src/main/java/net/minecraftforge/event/EntitySelectorEvent.java [hunk 5]
```diff
@@ -25,22 +25,17 @@
...
- * EntitySelectorEvent is fired whenever Minecraft collects entity selectors.
- * This happens (one or multiple times) when you use something like @a[gamemode=1] in a command.<br>
- * This event is fired via {@link ForgeEventFactory#gatherEntitySelectors(Map, String, ICommandSender, Vec3d)},
- * which is executed in {@link net.minecraft.command.EntitySelector#matchEntities(ICommandSender, String, Class)}<br>
- * <br>
- * This event is not cancelable and does not have a result.<br>
- * <br>
- * This event is fired on the {@link MinecraftForge#EVENT_BUS}
+ * Is not fired anymore.
+ * Replaced by a factory, which has to be registered with {@link net.minecraftforge.fml.common.registry.GameRegistry#registerEntitySelector(IEntitySelectorFactory, String...)}
```
Comment (mezz) [inline src/main/java/net/minecraftforge/event/EntitySelectorEvent.java:36]: Leave a javadoc comment saying it should be removed in 1.12

**Comment by mezz** [comment]
Ok, I'll take a look.

File: src/main/java/net/minecraftforge/fml/common/registry/GameRegistry.java [hunk 6]
```diff
@@ -95,6 +102,43 @@ public static void registerWorldGenerator(IWorldGenerator generator, int modGene
     }
 
     /**
+     * Registers a entity selector factory which is used to create predicates whenever a command containing selectors is executed
+     * Any non vanilla arguments that you expect has to be registered.
+     *
+     * @param arguments Expected string arguments in commands
```
Comment (mezz) [inline src/main/java/net/minecraftforge/fml/common/registry/GameRegistry.java]: I think this javadoc should be made more clear for people who have never implemented an entity selector before.
In the javadoc, please give a full example command that someone might type into the console, and point out which part of it is an "expected string argument".

File: src/main/java/net/minecraftforge/fml/common/registry/GameRegistry.java [hunk 7]
```diff
@@ -95,6 +102,43 @@ public static void registerWorldGenerator(IWorldGenerator generator, int modGene
     }
 
     /**
+     * Registers a entity selector factory which is used to create predicates whenever a command containing selectors is executed
+     * Any non vanilla arguments that you expect has to be registered.
```
Comment (mezz) [inline src/main/java/net/minecraftforge/fml/common/registry/GameRegistry.java]: I think it is important to point out *why* the arguments have to be registered. From what I understand, any argument that isn't added here will be ignored by the entity selector system.

File: src/main/java/net/minecraftforge/fml/common/IEntitySelectorFactory.java [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+import net.minecraft.entity.Entity;
+import net.minecraft.util.math.Vec3d;
+
+import javax.annotation.Nonnull;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Can be registered in {@link net.minecraftforge.fml.common.registry.GameRegistry} to generate entity predicates for custom entity selectors in commands.
```
Comment (mezz) [inline src/main/java/net/minecraftforge/fml/common/IEntitySelectorFactory.java]: This should contain the most important information first. It should also point directly to the registration method, GameRegistry is a large file to look through. Something like this:

```java
/**
 * Allows mods to create custom entity selectors in commands.
 * Registered in {@link net.minecraftforge.fml.common.registry.GameRegistry#registerEntitySelector}.
 * For an example implementation, see CustomEntitySelectorTest.
```

File: src/test/java/net/minecraftforge/test/CustomEntitySelectorTest.java [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+import net.minecraftforge.fml.common.event.FMLInitializationEvent;
+import net.minecraftforge.fml.common.registry.GameRegistry;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+@Mod(modid = "customentityselectortest", name = "Custom Entity Selector Test", version = "1.0", acceptableRemoteVersions = "*")
```
Comment (mezz) [inline src/test/java/net/minecraftforge/test/CustomEntitySelectorTest.java]: Please add a short javadoc with steps to use in-game to confirm that everything is working with this test.
Someone should be able to use this test to confirm everything is working, without reading the code. This also ensures that any bugs in the code can be checked against this javadoc specification if necessary.

**Review by mezz — CHANGES_REQUESTED** [review]
Looks good, I have some requests for changes in the documentation.

**Comment by maxanier** [comment]
Thank you very much for your feedback. I've updated the javadoc accordingly.

**Comment by mezz** [comment]
Tested and everything works as expected.

**Review by mezz — APPROVED** [review]

**Pull request merged**